### PR TITLE
open source companion to rstudio/rstudio-pro#3295

### PR DESCRIPTION
**Pull request by jgutman**
### Intent

There are a few issues which seem to be related to bugs or shortcomings in older versions of Boost (#7818, possibly #9920, #9854?). Updating RStudio bundled Boost to 1.78.0 relatively early in the cycle for Spotted Wakerobin so we have enough time for any issues to be uncovered.

### CODE REVIEW

**Comment by jgutman** [comment]
Note, the Boost tar has been uploaded to s3 but I still need to upload the windows binaries

File: src/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -297,13 +297,13 @@ else()
...
    endif()
    if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
-      set(BOOST_SUFFIX "vc141-mt-x${BOOST_ARCH}-1_69.lib")
-      set(BOOST_ROOT "${RSTUDIO_WINDOWS_DEPENDENCIES_DIR}/boost-1.69.0-win-msvc141-release-static/boost${BOOST_ARCH}")
+      set(BOOST_SUFFIX "vc141-mt-x${BOOST_ARCH}-1_78_0.lib")
+      set(BOOST_ROOT "${RSTUDIO_WINDOWS_DEPENDENCIES_DIR}/boost-1.78.0-win-msvc141-release-static/boost${BOOST_ARCH}")
    else()
-      set(BOOST_SUFFIX "vc141-mt-gd-x${BOOST_ARCH}-1_69.lib")
-      set(BOOST_ROOT "${RSTUDIO_WINDOWS_DEPENDENCIES_DIR}/boost-1.69.0-win-msvc141-debug-static/boost${BOOST_ARCH}")
+      set(BOOST_SUFFIX "vc141-mt-gd-x${BOOST_ARCH}-1_78_0.lib")
```
Comment (jgutman) [inline src/cpp/CMakeLists.txt:303]:
```suggestion
      set(BOOST_SUFFIX "vc141-mt-gd-x${BOOST_ARCH}-1_78.lib")
```

File: src/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -297,13 +297,13 @@ else()
...
+      set(BOOST_ROOT "${RSTUDIO_WINDOWS_DEPENDENCIES_DIR}/boost-1.78.0-win-msvc141-debug-static/boost${BOOST_ARCH}")
    endif()
-   set(BOOST_INCLUDEDIR "${BOOST_ROOT}/include/boost-1_69_0")
+   set(BOOST_INCLUDEDIR "${BOOST_ROOT}/include/boost-1_78_0")
```
Comment (jgutman) [inline src/cpp/CMakeLists.txt:306]:
```suggestion
   set(BOOST_INCLUDEDIR "${BOOST_ROOT}/include/boost-1_78")
```

File: src/cpp/CMakeLists.txt [hunk 3]
```diff
@@ -297,13 +297,13 @@ else()
       set(BOOST_ARCH "64")
    endif()
    if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
-      set(BOOST_SUFFIX "vc141-mt-x${BOOST_ARCH}-1_69.lib")
-      set(BOOST_ROOT "${RSTUDIO_WINDOWS_DEPENDENCIES_DIR}/boost-1.69.0-win-msvc141-release-static/boost${BOOST_ARCH}")
+      set(BOOST_SUFFIX "vc141-mt-x${BOOST_ARCH}-1_78_0.lib")
```
Comment (jgutman) [inline src/cpp/CMakeLists.txt:300]:
```suggestion
      set(BOOST_SUFFIX "vc141-mt-x${BOOST_ARCH}-1_78.lib")
```

File: dependencies/common/install-boost [hunk 4]
```diff
@@ -94,6 +95,7 @@ cd rstudio
 
 # bootstrap again
 ./bootstrap.sh
+cp b2 bjam
```
Comment (gtritchie) [inline dependencies/common/install-boost:98]: What does this do?

**Review by gtritchie — APPROVED** [review]
LGTM!

**Comment by jgutman** [comment]
Tested building with Boost 1.78 on RStudio Desktop on Mac OS 12.2.1 and Windows 11
On Mac this worked, on Windows I get the following when trying to build:
```
\Users\jacquelinegutman\rstudio\dependencies\windows\boost-1.78.0-win-msvc141-release-static\boost64\include\boost-1_78 \
   -IC:\Users\jacquelinegutman\rstudio\dependencies\windows\install-soci\sqlite\sqlite-amalgamation-3310100 -IC:\Users\jacquelinegutman\rstudio\dependencies\windows\install-soci\postgresql\include -IC:\Users\jacquelinegutman\rstudio\dependencies\windows\winpty-0.4.3-msys2-2.7.0\64\include -IC:\Users\jacquelinegutman\rstudio\dependencies\windows\openssl-1.1.1i\openssl-1.1.1i-release-64\include /DWIN32 /D_WINDOWS /W3 /GR /EHsc /std:c++14 /wd4800 /wd4091 /wd4068 /MP /bigobj /MD /O2 /Ob2 /DNDEBUG /showIncludes /Fosrc\cpp\core\CMakeFiles\rstudio-core.dir\system\Win32FileScanner.cpp.obj /Fdsrc\cpp\core\CMakeFiles\rstudio-core.dir\rstudio-core.pdb /FS \
   -c C:\Users\jacquelinegutman\rstudio\src\cpp\core\system\Win32FileScanner.cpp
C:\Users\jacquelinegutman\rstudio\src\cpp\core\system\Win32FileScanner.cpp(110): error C2065: '_1': undeclared identifier
C:\Users\jacquelinegutman\rstudio\src\cpp\core\system\Win32FileScanner.cpp(107): error C2672: \
   'std::transform': no matching overloaded function found
C:\Users\jacquelinegutman\rstudio\src\cpp\core\system\Win32FileScanner.cpp(110): error C2780: \
   '_OutIt std::transform(const _InIt1,const _InIt1,const _InIt2,_OutIt,_Fn)': expects 5 arguments - 3 provided
```

**Comment by gtritchie** [comment]
I don't know the gory details, but this sort of thing seems fairly common when upgrading boost. For example see similar tweaks needed when we went to boost 1.75, https://github.com/rstudio/rstudio/pulls?q=is%3Apr+is%3Aclosed+boost+in%3Atitle+.

It's possible you'll find additional places needing such tweaks in pro-repo-only code.

**Comment by jgutman** [comment]
> I don't know the gory details, but this sort of thing seems fairly common when upgrading boost. For example see similar tweaks needed when we went to boost 1.75, https://github.com/rstudio/rstudio/pulls?q=is%3Apr+is%3Aclosed+boost+in%3Atitle+.
> 
> It's possible you'll find additional places needing such tweaks in pro-repo-only code.

Makes sense, I explicitly namespaced the `_1` and that seems to fix the issue with building on Windows. I'll check for other places in the repo that might be affected. 

I've never actually built the Pro product, but I'll give a heads-up to the team

**Comment by gtritchie** [comment]
> I've never actually built the Pro product, but I'll give a heads-up to the team

Usually a change like this would be made in the pro repo, then the open-source subset pushed first: https://github.com/rstudio/rstudio-pro/wiki/Pro-and-Open-Source#reverse-merging.

The goal being to avoid breaking the pro builds.

**Review by kevinushey — COMMENTED** [review]
We could also consider using `#define BOOST_BIND_GLOBAL_PLACEHOLDERS` if we wanted to avoid requiring the `using namespace boost::placeholders` everywhere.

Given that we always use `boost::bind()` rather than `std::bind()`, I think that's reasonable.

**Comment by jgutman** [comment]
> We could also consider using `#define BOOST_BIND_GLOBAL_PLACEHOLDERS` if we wanted to avoid requiring the `using namespace boost::placeholders` everywhere.
> 
> Given that we always use `boost::bind()` rather than `std::bind()`, I think that's reasonable.

Ah smart! I didn't know about that. it was already in like a third to a half of our files, so I started adding it into the rest... but there's a lot. Although I'm probably 90% done at this point, but that seems like a reasonable idea

File: src/cpp/core/FileUtils.cpp [hunk 5]
```diff
@@ -29,7 +29,7 @@
 #include <core/system/PosixUser.hpp>
 #endif
 
-using namespace boost::placeholders;
+;
```
Comment (jgutman) [inline src/cpp/core/FileUtils.cpp:33]:
```suggestion
```

File: src/cpp/core/HtmlUtils.cpp [hunk 6]
```diff
@@ -27,7 +27,7 @@
 
 #include <core/http/Util.hpp>
 
-using namespace boost::placeholders;
+;
```
Comment (jgutman) [inline src/cpp/core/HtmlUtils.cpp:31]:
```suggestion
```

File: src/cpp/core/file_lock/FileLock.cpp [hunk 7]
```diff
@@ -32,7 +32,7 @@
 // borrowed from SessionConstants.hpp
 #define kRStudioSessionRoute "RSTUDIO_SESSION_ROUTE"
 
-using namespace boost::placeholders;
+;
```
Comment (jgutman) [inline src/cpp/core/file_lock/FileLock.cpp:36]:
```suggestion
```

File: src/cpp/core/gwt/GwtSymbolMaps.cpp [hunk 8]
```diff
@@ -30,7 +30,7 @@
 #include <shared_core/SafeConvert.hpp>
 #include <core/FileSerializer.hpp>
 
-using namespace boost::placeholders;
+;
```
Comment (jgutman) [inline src/cpp/core/gwt/GwtSymbolMaps.cpp:34]:
```suggestion
```

File: src/cpp/core/FileLockTests.cpp [hunk 9]
```diff
@@ -28,7 +28,7 @@
 #define RSTUDIO_NO_TESTTHAT_ALIASES
 #include <tests/TestThat.hpp>
 
-using namespace boost::placeholders;
+;
```
Comment (jgutman) [inline src/cpp/core/FileLockTests.cpp]:
```suggestion
```

File: src/cpp/core/FileUtils.cpp (hunk 5 above)
Comment (jgutman) [inline src/cpp/core/FileUtils.cpp]:
```suggestion
```

File: src/cpp/core/HtmlUtils.cpp (hunk 6 above)
Comment (jgutman) [inline src/cpp/core/HtmlUtils.cpp]:
```suggestion
```

File: src/cpp/core/file_lock/FileLock.cpp (hunk 7 above)
Comment (jgutman) [inline src/cpp/core/file_lock/FileLock.cpp]:
```suggestion
```

File: src/cpp/core/gwt/GwtSymbolMaps.cpp (hunk 8 above)
Comment (jgutman) [inline src/cpp/core/gwt/GwtSymbolMaps.cpp]:
```suggestion
```

File: src/cpp/core/include/core/http/StreamWriter.hpp [hunk 10]
```diff
@@ -22,6 +22,8 @@
 #include <core/http/Response.hpp>
 #include <core/http/AsyncClient.hpp>
 
+;
+
```
Comment (jgutman) [inline src/cpp/core/include/core/http/StreamWriter.hpp:26]:
```suggestion
```

File: src/cpp/core/system/Win32FileScanner.cpp [hunk 11]
```diff
@@ -24,6 +24,8 @@
 #include <shared_core/Error.hpp>
 #include <shared_core/FilePath.hpp>
 
+
+
```
Comment (jgutman) [inline src/cpp/core/system/Win32FileScanner.cpp:28]:
```suggestion
```

File: src/cpp/session/modules/SessionConsole.cpp [hunk 12]
```diff
@@ -38,6 +38,7 @@
 #define kMinConsoleLines 10
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/SessionConsole.cpp:41]:
```suggestion
```

File: src/cpp/session/modules/SessionPlots.cpp [hunk 13]
```diff
@@ -42,6 +42,7 @@
 #include <session/SessionModuleContext.hpp>
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/SessionPlots.cpp:45]:
```suggestion
```

File: src/cpp/session/modules/SessionShinyViewer.cpp [hunk 14]
```diff
@@ -41,6 +41,7 @@
 #define kForegroudAppId "foreground"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/SessionShinyViewer.cpp:44]:
```suggestion
```

File: src/cpp/session/modules/build/SessionBuild.cpp [hunk 15]
```diff
@@ -59,6 +59,7 @@
 #include "SessionInstallRtools.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/build/SessionBuild.cpp:62]:
```suggestion
```

File: src/cpp/session/modules/clang/DefinitionIndex.cpp [hunk 16]
```diff
@@ -34,6 +34,7 @@
 
 using namespace rstudio::core;
 using namespace rstudio::core::libclang;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/clang/DefinitionIndex.cpp:37]:
```suggestion
```

File: src/cpp/session/modules/clang/RCompilationDatabase.cpp [hunk 17]
```diff
@@ -48,6 +48,7 @@
 
 using namespace rstudio::core;
 using namespace rstudio::core::libclang;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/clang/RCompilationDatabase.cpp:51]:
```suggestion
```

File: src/cpp/session/modules/connections/ConnectionHistory.cpp [hunk 18]
```diff
@@ -21,6 +21,7 @@
 #include <session/SessionModuleContext.hpp>
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/connections/ConnectionHistory.cpp:24]:
```suggestion
```

**Comment by netique** [comment]
Maybe OT a bit, but are you able to build RStudio on Windows with Boost 1.69.0? I get error that Boost cannot be found no matter how i set env vars (or not, my first attempts were following dependencies scripts from the repo). The log says Boost_INCLUDE_DIR is missing.

File: src/cpp/session/modules/environment/EnvironmentMonitor.cpp [hunk 19]
```diff
@@ -23,6 +23,7 @@
 #include "EnvironmentUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/environment/EnvironmentMonitor.cpp:26]:
```suggestion
```

File: src/cpp/session/modules/environment/SessionEnvironment.cpp [hunk 20]
```diff
@@ -38,6 +38,7 @@
 #include "EnvironmentUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/environment/SessionEnvironment.cpp:41]:
```suggestion
```

File: src/cpp/session/modules/panmirror/SessionPanmirrorDOI.cpp [hunk 21]
```diff
@@ -30,6 +30,7 @@
 #include "SessionPanmirrorUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/panmirror/SessionPanmirrorDOI.cpp:33]:
```suggestion
```

File: src/cpp/session/modules/panmirror/SessionPanmirrorDataCite.cpp [hunk 22]
```diff
@@ -28,6 +28,7 @@
 #include "SessionPanmirrorUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/panmirror/SessionPanmirrorDataCite.cpp:31]:
```suggestion
```

File: src/cpp/session/projects/SessionProjectContext.cpp [hunk 23]
```diff
@@ -46,6 +46,7 @@
 #define kStorageFolder "projects"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/projects/SessionProjectContext.cpp:49]:
```suggestion
```

File: src/cpp/session/prefs/PrefLayer.cpp [hunk 24]
```diff
@@ -18,7 +18,7 @@
 #include <core/FileSerializer.hpp>
 #include <core/Algorithm.hpp>
 
-using namespace rstudio::core;
+using namespace rstudio::core
```
Comment (jgutman) [inline src/cpp/session/prefs/PrefLayer.cpp:22]:
```suggestion
using namespace rstudio::core;

```

File: src/cpp/session/modules/zotero/ZoteroCollectionsWeb.cpp [hunk 25]
```diff
@@ -29,6 +29,7 @@
 #include "ZoteroCSL.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/zotero/ZoteroCollectionsWeb.cpp:32]:
```suggestion
```

File: src/cpp/session/modules/zotero/SessionZotero.cpp [hunk 26]
```diff
@@ -34,6 +34,7 @@
 #include "ZoteroUtil.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/zotero/SessionZotero.cpp:37]:
```suggestion
```

File: src/cpp/session/modules/tex/SessionRnwWeave.cpp [hunk 27]
```diff
@@ -40,6 +40,7 @@
 
 using namespace rstudio::core;
 using namespace rstudio::session::modules::tex::rnw_concordance;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/tex/SessionRnwWeave.cpp:43]:
```suggestion
```

File: src/cpp/session/modules/tex/SessionRnwConcordance.cpp [hunk 28]
```diff
@@ -35,6 +35,7 @@
 #include <session/SessionModuleContext.hpp>
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/tex/SessionRnwConcordance.cpp:38]:
```suggestion
```

File: src/cpp/session/modules/tex/SessionPdfLatex.cpp [hunk 29]
```diff
@@ -31,6 +31,7 @@
 #include "SessionTexUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/tex/SessionPdfLatex.cpp:34]:
```suggestion
```

File: src/cpp/session/modules/tex/SessionCompilePdfSupervisor.cpp [hunk 30]
```diff
@@ -23,6 +23,7 @@
 #include <session/SessionModuleContext.hpp>
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/tex/SessionCompilePdfSupervisor.cpp:26]:
```suggestion
```

File: src/cpp/session/modules/tex/SessionCompilePdf.cpp [hunk 31]
```diff
@@ -47,6 +47,7 @@
 #include "SessionTexUtils.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/tex/SessionCompilePdf.cpp:50]:
```suggestion
```

File: src/cpp/session/modules/rmarkdown/SessionBookdown.cpp [hunk 32]
```diff
@@ -26,6 +26,7 @@
 #include "SessionBookdownXRefs.hpp"
 
 using namespace rstudio::core;
+using namespace boost::placeholders;
```
Comment (jgutman) [inline src/cpp/session/modules/rmarkdown/SessionBookdown.cpp:29]:
```suggestion
```

**Comment by jgutman** [comment]
> Maybe OT a bit, but are you able to build RStudio on Windows with Boost 1.69.0? I get error that Boost cannot be found no matter how i set env vars (or not, my first attempts were following dependencies scripts from the repo). The log says Boost_INCLUDE_DIR is missing.

Yes, we've been building RStudio on Windows with Boost 1.69, that is the bundled version we currently use

**Comment by jgutman** [comment]
Confirmed development builds work with open source RStudio Desktop on Mac OS and Windows, and open source RStudio Server.

**Pull request merged**